libtransact/src/context/manager.rs

```diff
@@ -0,0 +1,528 @@
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::error::Error as StdError;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::str;
+
+use crate::context::execution_context::Context;
+use crate::context::ContextId;
+use crate::receipts::{Event, StateChange, TransactionReceipt, TransactionReceiptBuilder};
+use crate::state::Read;
+
+/// Enum to hold the pertinent information based on the changes made by a StateChange.
+/// A StateChange::Set returns a StateChangeResult::Set with the Key for the Value altered.
+/// A StateChange::Delete returns a StateChangeResult::Delete with the deleted Value.
+#[derive(Debug)]
+pub enum StateChangeResult<K, V> {
+    Set { key: K },
+    Delete { value: V },
+}
+
+#[derive(Debug)]
+pub enum ContextManagerError {
+    MissingContextError(String),
+}
+
+impl StdError for ContextManagerError {
+    fn description(&self) -> &str {
+        match *self {
+            ContextManagerError::MissingContextError(ref msg) => msg,
+        }
+    }
+}
+
+impl std::fmt::Display for ContextManagerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ContextManagerError::MissingContextError(ref s) => {
+                write!(f, "Unable to find specified Context: {:?}", s)
+            }
+        }
+    }
+}
+
+pub struct ContextManager<K, V, R: Read<StateId = String, Key = K, Value = V>> {
+    contexts: HashMap<ContextId, Context<K, V>>,
+    database: R,
+}
+impl<
+        K: Hash + Eq + Clone + Debug + Default,
+        V: Clone + Debug + Default,
+        R: Read<StateId = String, Key = K, Value = V>,
+    > ContextManager<K, V, R>
+{
+    pub fn new(database: R) -> Self {
+        ContextManager {
+            contexts: HashMap::new(),
+            database,
+        }
+    }
+
+    /// Returns a mutable Context within the ContextManager's Context list specified by the ContextId
+    fn get_context_mut(
+        &mut self,
+        context_id: &ContextId,
+    ) -> Result<&mut Context<K, V>, ContextManagerError> {
+        self.contexts.get_mut(context_id).ok_or_else(|| {
+            ContextManagerError::MissingContextError(
+                str::from_utf8(context_id)
+                    .expect("Unable to generate string from ContextId")
+                    .to_string(),
+            )
+        })
+    }
+
+    /// Returns a Context within the ContextManager's Context list specified by the ContextId
+    fn get_context(&self, context_id: &ContextId) -> Result<Context<K, V>, ContextManagerError> {
```

Comment:
This probably can return a reference to the context, so as not to clone the underlying context.

Reply:
Addressed